File: java/agentbay/src/main/java/com/aliyun/agentbay/filesystem/FileSystem.java
package com.aliyun.agentbay.filesystem;

import com.aliyun.agentbay.exception.AgentBayException;
import com.aliyun.agentbay.exception.FileException;
import com.aliyun.agentbay.model.*;
import com.aliyun.agentbay.service.BaseService;
import com.aliyun.agentbay.session.Session;
import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;
import java.util.*;

/**
 * File system operations for a session
 */
public class FileSystem extends BaseService {
    private static final ObjectMapper objectMapper = new ObjectMapper();

    private static final int DEFAULT_CHUNK_SIZE = 50 * 1024;
    private static final int DEFAULT_VPC_CHUNK_SIZE = 512 * 1024;

    private FileTransfer fileTransfer;

    public FileSystem(Session session) {
        super(session);
    }

    private FileTransfer ensureFileTransfer() {
        if (fileTransfer == null) {
            if (session.getAgentBay() == null) {
                throw new RuntimeException("FileTransfer requires an AgentBay instance");
            }
            if (session == null) {
                throw new RuntimeException("FileTransfer requires a session");
            }
            fileTransfer = new FileTransfer(session.getAgentBay(), session);
        }
        return fileTransfer;
    }

    public String getFileTransferContextPath() {
        FileTransfer transfer = ensureFileTransfer();
        return transfer.getContextPath();
    }

    /**
     * Read a file
     *
     * @param path File path
     * @return File content
     * @throws AgentBayException if reading fails
     */
    public String read(String path) throws AgentBayException {
        Map<String, Object> args = new HashMap<>();
        args.put("path", path);
        OperationResult result = callMcpTool("read_file", args);
        if (result.isSuccess()) {
            return result.getData();
        } else {
            throw new AgentBayException(result.getErrorMessage());
        }
    }

    /**
     * Write content to a file
     *
     * @param path File path
     * @param content Content to write
     * @return Write result
     * @throws AgentBayException if writing fails
     */
    public String write(String path, String content) throws AgentBayException {
        Map<String, Object> args = new HashMap<>();
        args.put("path", path);
        args.put("content", content);
        OperationResult result = callMcpTool("write_file", args);
        if (result.isSuccess()) {
            return result.getData();
        } else {
            throw new AgentBayException(result.getErrorMessage());
        }
    }

    /**
     * List directory contents
     *
     * @param path Directory path
     * @return Directory listing
     * @throws AgentBayException if listing fails
     */
    public String list(String path) throws AgentBayException {
        Map<String, Object> args = new HashMap<>();
        args.put("path", path);
        OperationResult result = callMcpTool("list_directory", args);
        if (result.isSuccess()) {
            return result.getData();
        } else {
            throw new AgentBayException(result.getErrorMessage());
        }
    }

    /**
     * Check if a file or directory exists
     *
     * @param path Path to check
     * @return true if exists, false otherwise
     * @throws AgentBayException if check fails
     */
    public boolean exists(String path) throws AgentBayException {
        try {
            Map<String, Object> args = new HashMap<>();
            args.put("command", "test -e \"" + path + "\" && echo 'exists' || echo 'not exists'");
            OperationResult result = callMcpTool("shell", args);
            return result.isSuccess() && result.getData() != null && result.getData().trim().equals("exists");
        } catch (Exception e) {
            return false;
        }
    }

    /**
     * Create a directory
     *
     * @param path Directory path
     * @return Creation result
     * @throws AgentBayException if creation fails
     */
    public String mkdir(String path) throws AgentBayException {
        Map<String, Object> args = new HashMap<>();
        args.put("command", "mkdir -p \"" + path + "\"");
        OperationResult result = callMcpTool("shell", args);
        if (result.isSuccess()) {
            return result.getData();
        } else {
            throw new AgentBayException(result.getErrorMessage());
        }
    }

    /**
     * Remove a file or directory using shell command
     *
     * @param path Path to remove
     * @return Removal result
     * @throws AgentBayException if removal fails
     * @deprecated Use {@link #deleteFile(String)} instead. This method uses shell command which may not work in all environments.
     */
    @Deprecated
    public String removeLegacy(String path) throws AgentBayException {
        Map<String, Object> args = new HashMap<>();
        args.put("command", "rm -rf \"" + path + "\"");
        OperationResult result = callMcpTool("shell", args);
        if (result.isSuccess()) {
            return result.getData();
        } else {
            throw new AgentBayException(result.getErrorMessage());
        }
    }

    /**
     * Copy a file or directory
     *
     * @param source Source path
     * @param destination Destination path
     * @return Copy result
     * @throws AgentBayException if copy fails
     */
    public String copy(String source, String destination) throws AgentBayException {
        Map<String, Object> args = new HashMap<>();
        args.put("command", "cp -r \"" + source + "\" \"" + destination + "\"");
        OperationResult result = callMcpTool("shell", args);
        if (result.isSuccess()) {
            return result.getData();
        } else {
            throw new AgentBayException(result.getErrorMessage());
        }
    }

    /**
     * Move a file or directory
     *
     * @param source Source path
     * @param destination Destination path
     * @return Move result
     * @throws AgentBayException if move fails
     */
    public String move(String source, String destination) throws AgentBayException {
        Map<String, Object> args = new HashMap<>();
        args.put("command", "mv \"" + source + "\" \"" + destination + "\"");
        OperationResult result = callMcpTool("shell", args);
        if (result.isSuccess()) {
            return result.getData();
        } else {
            throw new AgentBayException(result.getErrorMessage());
        }
    }

    /**
     * Get file information
     *
     * @param path File path
     * @return File information
     * @throws AgentBayException if getting info fails
     */
    public String getInfo(String path) throws AgentBayException {
        Map<String, Object> args = new HashMap<>();
        args.put("command", "ls -la \"" + path + "\"");
        OperationResult result = callMcpTool("shell", args);
        if (result.isSuccess()) {
            return result.getData();
        } else {
            throw new AgentBayException(result.getErrorMessage());
        }
    }

    /**
     * Write content to a file. Automatically handles large files by chunking.
     * Similar to Python's write_file method.
     *
     * @param path File path
     * @param content Content to write
     * @param mode Write mode ("overwrite" or "append")
     * @param createParentDir Whether to create parent directories if they don't exist (default: false)
     * @return BoolResult containing success status and error message if any
     */
    public BoolResult writeFile(String path, String content, String mode, boolean createParentDir) {
        int contentLength = content.getBytes().length;
        int chunkSize = DEFAULT_CHUNK_SIZE;
        if (session.isVpcEnabled()) {
            chunkSize =DEFAULT_VPC_CHUNK_SIZE;
        }
        // If the content length is less than the chunk size, write it directly
        if (contentLength <= chunkSize) {
            return writeFileChunk(path, content, mode, createParentDir);
        }

        try {
            // Write the first chunk (creates or overwrites the file)
            String firstChunk = content.substring(0, Math.min(chunkSize, content.length()));
            BoolResult result = writeFileChunk(path, firstChunk, mode, createParentDir);
            if (!result.isSuccess()) {
                return result;
            }

            // Write the rest in chunks (appending)
            int offset = chunkSize;
            while (offset < content.length()) {
                int end = Math.min(offset + chunkSize, content.length());
                String currentChunk = content.substring(offset, end);
                result = writeFileChunk(path, currentChunk, "append", false);
                if (!result.isSuccess()) {
                    return result;
                }
                offset = end;
            }

            return new BoolResult(result.getRequestId(), true, true, "");

        } catch (Exception e) {
            return new BoolResult("", false, false, "Failed to write file: " + e.getMessage());
        }
    }

    /**
     * Write file content using BoolResult with default "overwrite" mode
     */
    public BoolResult writeFile(String path, String content) {
        return writeFile(path, content, "overwrite", false);
    }

    /**
     * Write file content using BoolResult with specified mode
     */
    public BoolResult writeFile(String path, String content, String mode) {
        return writeFile(path, content, mode, false);
    }

    // Note: The three-parameter version writeFile(path, content, mode) is the main method above

    /**
     * Internal method to write a file chunk. Used for chunked file operations.
     * Similar to Python's _write_file_chunk method.
     *
     * @param path File path
     * @param content Content to write
     * @param mode Write mode ("overwrite" or "append")
     * @param createParentDir Whether to create parent directories if they don't exist
     * @return BoolResult containing success status and error message if any
     */
    private BoolResult writeFileChunk(String path, String content, String mode, boolean createParentDir) {
        if (!"overwrite".equals(mode) && !"append".equals(mode)) {
            return new BoolResult("", false, false,
                "Invalid write mode: " + mode + ". Must be 'overwrite' or 'append'.");
        }

        Map<String, Object> args = new HashMap<>();
        args.put("path", path);
        args.put("content", content);
        args.put("mode", mode);
        args.put("create_parent_dir", createParentDir);

        try {
            OperationResult result = callMcpTool("write_file", args);
            if (result.isSuccess()) {
                return new BoolResult(result.getRequestId(), true, true, "");
            } else {
                return new BoolResult(result.getRequestId(), false, false, result.getErrorMessage());
            }

        } catch (Exception e) {
            return new BoolResult("", false, false, "Failed to write file chunk: " + e.getMessage());
        }
    }

    /**
     * Read file content using FileContentResult
     */
    public FileContentResult readFile(String path) {
        Map<String, Object> args = new HashMap<>();
        args.put("path", path);

        try {
            OperationResult result = callMcpTool("read_file", args);
            if (result.isSuccess()) {
                return new FileContentResult(result.getRequestId(), true, result.getData(), "");
            } else {
                return new FileContentResult(result.getRequestId(), false, "", result.getErrorMessage());
            }

        } catch (Exception e) {
            return new FileContentResult("", false, "", "Failed to read file: " + e.getMessage());
        }
    }

    /**
     * Search for files matching a pattern
     */
    public FileSearchResult searchFiles(String directory, String pattern) {
        return searchFiles(directory, pattern, null);
    }

    public FileSearchResult searchFiles(String directory, String pattern, List<String> excludePatterns) {
        try {
            Map<String, Object> args = new HashMap<>();
            args.put("path", directory);
            args.put("pattern", pattern);
            if (excludePatterns != null && !excludePatterns.isEmpty()) {
                args.put("excludePatterns", excludePatterns);
            }

            OperationResult result = callMcpTool("search_files", args);

            if (result.isSuccess()) {
                String[] matchingFiles = result.getData() != null && !result.getData().trim().isEmpty()
                    ? result.getData().trim().split("\n")
                    : new String[0];

                List<String> matches = Arrays.asList(matchingFiles);
                if (matches.size() == 1 && "No matches found".equals(matches.get(0))) {
                    return new FileSearchResult(result.getRequestId(), false, new ArrayList<>(), "No matches found");
                }
                return new FileSearchResult(result.getRequestId(), true, matches, "");
            } else {
                return new FileSearchResult(result.getRequestId(), false, null,
                    result.getErrorMessage() != null ? result.getErrorMessage() : "Failed to search files");
            }
        } catch (Exception e) {
            return new FileSearchResult("", false, null, e.getMessage());
        }
    }

    /**
     * Read multiple files at once
     */
    public MultipleFileContentResult readMultipleFiles(List<String> paths) {
        try {
            Map<String, Object> args = new HashMap<>();
            args.put("paths", paths);

            OperationResult result = callMcpTool("read_multiple_files", args);

            if (result.isSuccess()) {
                Map<String, String> filesContent = parseMultipleFilesResponse(result.getData());
                return new MultipleFileContentResult(result.getRequestId(), true, filesContent, "");
            } else {
                return new MultipleFileContentResult(result.getRequestId(), false, null,
                    result.getErrorMessage() != null ? result.getErrorMessage() : "Failed to read multiple files");
            }
        } catch (Exception e) {
            return new MultipleFileContentResult("", false, null, e.getMessage());
        }
    }

    /**
     * Delete a file at the specified path.
     *
     * @param path The path of the file to delete
     * @return BoolResult containing success status and error message if any
     *
     * Example:
     * <pre>
     * AgentBay agentBay = new AgentBay(System.getenv("AGENTBAY_API_KEY"));
     * SessionResult result = agentBay.create(new CreateSessionParams());
     * Session session = result.getSession();
     * session.getFileSystem().writeFile("/tmp/to_delete.txt", "hello");
     * BoolResult deleteResult = session.getFileSystem().deleteFile("/tmp/to_delete.txt");
     * session.delete();
     * </pre>
     */
    public BoolResult  deleteFile(String path) {
        Map<String, Object> args = new HashMap<>();
        args.put("path", path);

        try {
            OperationResult result = callMcpTool("delete_file", args);
            if (result.isSuccess()) {
                return new BoolResult(result.getRequestId(), true, true, "");
            } else {
                return new BoolResult(result.getRequestId(), false, false, result.getErrorMessage());
            }

        } catch (Exception e) {
            return new BoolResult("", false, false, "Failed to delete file: " + e.getMessage());
        }
    }

    public BoolResult createDirectory(String path) {
        Map<String, Object> args = new HashMap<>();
        args.put("path", path);

        try {
            OperationResult result = callMcpTool("create_directory", args);
            if (result.isSuccess()) {
                return new BoolResult(result.getRequestId(), true, true, "");
            } else {
                return new BoolResult(result.getRequestId(), false, false, result.getErrorMessage());
            }

        } catch (Exception e) {
            return new BoolResult("", false, false, "Failed to create directory: " + e.getMessage());
        }
    }

    public com.aliyun.agentbay.model.DirectoryListResult listDirectory(String path) {
        try {
            Map<String, Object> args = new HashMap<>();
            args.put("path", path);

            OperationResult result = callMcpTool("list_directory", args);

            if (result.isSuccess()) {
                // Parse the directory listing response like Python SDK does
                String directoryResponse = result.getData();
                List<Map<String, Object>> entries = parseDirectoryListing(directoryResponse);

                return new com.aliyun.agentbay.model.DirectoryListResult(result.getRequestId(), true, entries, "");
            } else {
                return new com.aliyun.agentbay.model.DirectoryListResult(result.getRequestId(), false, null, result.getErrorMessage());
            }
        } catch (Exception e) {
            return new com.aliyun.agentbay.model.DirectoryListResult("", false, null, e.getMessage());
        }
    }

    public com.aliyun.agentbay.model.FileInfoResult getFileInfo(String path) {
        try {
            Map<String, Object> args = new HashMap<>();
            args.put("path", path);

            OperationResult result = callMcpTool("get_file_info", args);

            if (result.isSuccess()) {
                Map<String, Object> fileInfo = parseFileInfo(result.getData());
                com.aliyun.agentbay.model.FileInfoResult fileInfoResult =
                    new com.aliyun.agentbay.model.FileInfoResult(result.getRequestId(), true, result.getData(), "");
                fileInfoResult.setFileInfo(fileInfo);
                return fileInfoResult;
            } else {
                return new com.aliyun.agentbay.model.FileInfoResult(result.getRequestId(), false, "",
                    result.getErrorMessage() != null ? result.getErrorMessage() : "Failed to get file info");
            }
        } catch (Exception e) {
            return new com.aliyun.agentbay.model.FileInfoResult("", false, null, e.getMessage());
        }
    }

    public BoolResult editFile(String path, java.util.List<java.util.Map<String, String>> edits) {
        return editFile(path, edits, false);
    }

    public BoolResult editFile(String path, java.util.List<java.util.Map<String, String>> edits, boolean dryRun) {
        Map<String, Object> args = new HashMap<>();
        args.put("path", path);
        args.put("edits", edits);
        args.put("dryRun", dryRun);

        try {
            OperationResult result = callMcpTool("edit_file", args);
            if (result.isSuccess()) {
                return new BoolResult(result.getRequestId(), true, true, "");
            } else {
                return new BoolResult(result.getRequestId(), false, false, result.getErrorMessage());
            }

        } catch (Exception e) {
            return new BoolResult("", false, false, "Failed to edit file: " + e.getMessage());
        }
    }

    public BoolResult moveFile(String source, String destination) {
        try {
            Map<String, Object> args = new HashMap<>();
            args.put("source", source);
            args.put("destination", destination);

            OperationResult result = callMcpTool("move_file", args);

            if (result.isSuccess()) {
                return new BoolResult(result.getRequestId(), true, true, "");
            } else {
                return new BoolResult(result.getRequestId(), false, false,
                    result.getErrorMessage() != null ? result.getErrorMessage() : "Failed to move file");
            }
        } catch (Exception e) {
            return new BoolResult("", false, false, e.getMessage());
        }
    }

    /**
     * Parse directory listing text into a list of file/directory entries.
     * This mimics the Python SDK's parse_directory_listing function.
     *
     * @param text Directory listing text in format:
     *             [DIR] directory_name
     *             [FILE] file_name
     *             Each entry should be on a new line with [DIR] or [FILE] prefix
     * @return List of maps containing name and isDirectory fields
     */
    private List<Map<String, Object>> parseDirectoryListing(String text) {
        List<Map<String, Object>> result = new ArrayList<>();
        if (text == null || text.trim().isEmpty()) {
            return result;
        }

        String[] lines = text.split("\n");
        for (String line : lines) {
            line = line.trim();
            if (line.isEmpty()) {
                continue;
            }

            Map<String, Object> entryMap = new HashMap<>();
            if (line.startsWith("[DIR]")) {
                entryMap.put("isDirectory", true);
                entryMap.put("name", line.replace("[DIR]", "").trim());
            } else if (line.startsWith("[FILE]")) {
                entryMap.put("isDirectory", false);
                entryMap.put("name", line.replace("[FILE]", "").trim());
            } else {
                // Skip lines that don't match the expected format
                continue;
            }

            result.add(entryMap);
        }

        return result;
    }

    /**
     * Parse file info string into a map.
     * This mimics the Python SDK's parse_file_info function.
     *
     * @param fileInfoStr The file info string to parse
     * @return A map containing the parsed file info
     */
    private Map<String, Object> parseFileInfo(String fileInfoStr) {
        Map<String, Object> result = new HashMap<>();
        if (fileInfoStr == null || fileInfoStr.trim().isEmpty()) {
            return result;
        }

        String[] lines = fileInfoStr.split("\\n");
        for (String line : lines) {
            if (line.contains(":")) {
                String[] parts = line.split(":", 2);
                if (parts.length == 2) {
                    String key = parts[0].trim();
                    String value = parts[1].trim();

                    // Convert boolean values
                    if ("true".equalsIgnoreCase(value)) {
                        result.put(key, true);
                    } else if ("false".equalsIgnoreCase(value)) {
                        result.put(key, false);
                    } else {
                        // Try to convert numeric values
                        try {
                            if (value.contains(".")) {
                                result.put(key, Double.parseDouble(value));
                            } else {
                                result.put(key, Long.parseLong(value));
                            }
                        } catch (NumberFormatException e) {
                            // Keep as string if not numeric
                            result.put(key, value);
                        }
                    }
                }
            }
        }
        return result;
    }
    /**
     * Parse the response from reading multiple files.
     * This mimics the Python SDK's parse_multiple_files_response function.
     *
     * @param text The response string containing file contents in format:
     *             /path/to/file1.txt: Content of file1
     *
     *             ---
     *
     *             /path/to/file2.txt:
     *             Content of file2
     * @return A map mapping file paths to their content
     */
    private Map<String, String> parseMultipleFilesResponse(String text) {
        Map<String, String> result = new HashMap<>();
        if (text == null || text.trim().isEmpty()) {
            return result;
        }

        String[] lines = text.split("\\n");
        String currentPath = null;
        List<String> currentContent = new ArrayList<>();

        for (int i = 0; i < lines.length; i++) {
            String line = lines[i];

            // Check if this line contains a file path (ends with a colon)
            if (line.contains(":") && currentPath == null) {
                // Extract path (everything before the first colon)
                int colonIndex = line.indexOf(":");
                String path = line.substring(0, colonIndex).trim();

                // Start collecting content (everything after the colon)
                currentPath = path;

                // If there's content on the same line after the colon, add it
                if (line.length() > colonIndex + 1) {
                    String contentStart = line.substring(colonIndex + 1).trim();
                    if (!contentStart.isEmpty()) {
                        currentContent.add(contentStart);
                    }
                }
            }
            // Check if this is a separator line
            else if ("---".equals(line.trim())) {
                // Save the current file content
                if (currentPath != null) {
                    result.put(currentPath, String.join("\\n", currentContent).trim());
                    currentPath = null;
                    currentContent.clear();
                }
            }
            // If we're collecting content for a path, add this line
            else if (currentPath != null) {
                currentContent.add(line);
            }
        }

        // Save the last file content if exists
        if (currentPath != null) {
            result.put(currentPath, String.join("\\n", currentContent).trim());
        }

        return result;
    }

    public UploadResult uploadFile(
        String localPath,
        String remotePath,
        String contentType,
        boolean wait,
        float waitTimeout,
        float pollInterval
    ) {
        return uploadFile(localPath, remotePath, contentType, wait, waitTimeout, pollInterval, null);
    }

    public UploadResult uploadFile(
        String localPath,
        String remotePath,
        String contentType,
        boolean wait,
        float waitTimeout,
        float pollInterval,
        ProgressCallback progressCallback
    ) {
        try {
            FileTransfer transfer = ensureFileTransfer();
            UploadResult result = transfer.upload(localPath, remotePath, contentType, wait, waitTimeout, pollInterval, progressCallback);
            return result;
        } catch (Exception e) {
            return new UploadResult(
                "", false, null, null, null, null, 0, remotePath,
                "Upload failed: " + e.getMessage()
            );
        }
    }

    public UploadResult uploadFile(String localPath, String remotePath) {
        return uploadFile(localPath, remotePath, null, true, 30.0f, 1.5f);
    }

    public DownloadResult downloadFile(
        String remotePath,
        String localPath,
        boolean overwrite,
        boolean wait,
        float waitTimeout,
        float pollInterval
    ) {
        return downloadFile(remotePath, localPath, overwrite, wait, waitTimeout, pollInterval, null);
    }

    public DownloadResult downloadFile(
        String remotePath,
        String localPath,
        boolean overwrite,
        boolean wait,
        float waitTimeout,
        float pollInterval,
        ProgressCallback progressCallback
    ) {
        try {
            FileTransfer transfer = ensureFileTransfer();
            DownloadResult result = transfer.download(remotePath, localPath, overwrite, wait, waitTimeout, pollInterval, progressCallback);
            return result;
        } catch (Exception e) {
            return new DownloadResult(
                "", false, null, null, null, 0, remotePath, localPath, null,
                "Download failed: " + e.getMessage()
            );
        }
    }

    public DownloadResult downloadFile(String remotePath, String localPath) {
        return downloadFile(remotePath, localPath, true, true, 30.0f, 1.5f);
    }

    public DownloadResult downloadFile(String remotePath) {
        return downloadFile(remotePath, null);
    }

    /**
     * Upload byte array to remote path using pre-signed URLs.
     * This is a Java SDK extension that allows uploading data from memory without writing to disk first.
     *
     * @param content Byte array content to upload
     * @param remotePath Remote file path to upload to
     * @param contentType Optional content type for the file
     * @param wait Whether to wait for the sync operation to complete
     * @param waitTimeout Timeout for waiting for sync completion (seconds)
     * @param pollInterval Interval between polling for sync completion (seconds)
     * @param progressCallback Callback for upload progress updates
     * @return UploadResult containing the result of the upload operation
     */
    public UploadResult uploadFileBytes(
        byte[] content,
        String remotePath,
        String contentType,
        boolean wait,
        float waitTimeout,
        float pollInterval,
        ProgressCallback progressCallback
    ) {
        try {
            FileTransfer transfer = ensureFileTransfer();
            UploadResult result = transfer.uploadBytes(content, remotePath, contentType, wait, waitTimeout, pollInterval, progressCallback);
            return result;
        } catch (Exception e) {
            return new UploadResult(
                "", false, null, null, null, null, 0, remotePath,
                "Upload failed: " + e.getMessage()
            );
        }
    }

    /**
     * Upload byte array to remote path with default parameters.
     *
     * @param content Byte array content to upload
     * @param remotePath Remote file path to upload to
     * @return UploadResult containing the result of the upload operation
     */
    public UploadResult uploadFileBytes(byte[] content, String remotePath) {
        return uploadFileBytes(content, remotePath, null, true, 30.0f, 1.5f, null);
    }

    /**
     * Download file from remote path to byte array using pre-signed URLs.
     * This is a Java SDK extension that allows downloading data to memory without writing to disk.
     *
     * @param remotePath Remote file path to download from
     * @param wait Whether to wait for the sync operation to complete
     * @param waitTimeout Timeout for waiting for sync completion (seconds)
     * @param pollInterval Interval between polling for sync completion (seconds)
     * @param progressCallback Callback for download progress updates
     * @return DownloadResult containing the downloaded byte array in the content field
     */
    public DownloadResult downloadFileBytes(
        String remotePath,
        boolean wait,
        float waitTimeout,
        float pollInterval,
        ProgressCallback progressCallback
    ) {
        try {
            FileTransfer transfer = ensureFileTransfer();
            DownloadResult result = transfer.downloadBytes(remotePath, wait, waitTimeout, pollInterval, progressCallback);
            return result;
        } catch (Exception e) {
            return new DownloadResult(
                "", false, null, null, null, 0, remotePath, null, null,
                "Download failed: " + e.getMessage()
            );
        }
    }

    /**
     * Download file from remote path to byte array with default parameters.
     *
     * @param remotePath Remote file path to download from
     * @return DownloadResult containing the downloaded byte array in the content field
     */
    public DownloadResult downloadFileBytes(String remotePath) {
        return downloadFileBytes(remotePath, true, 30.0f, 1.5f, null);
    }

    public BoolResult delete(String path) {
        return deleteFile(path);
    }

    public BoolResult remove(String path) {
        return deleteFile(path);
    }

    public BoolResult rm(String path) {
        return deleteFile(path);
    }

    public com.aliyun.agentbay.model.DirectoryListResult ls(String path) {
        return listDirectory(path);
    }
}

File: python/agentbay/_common/models/filesystem.py
"""
Filesystem module data models.
"""

from dataclasses import dataclass
from typing import Any, Dict, List, Optional, Union

from .response import ApiResponse


@dataclass
class UploadResult:
    """Result structure for file upload operations."""

    success: bool
    request_id_upload_url: Optional[str]
    request_id_sync: Optional[str]
    http_status: Optional[int]
    etag: Optional[str]
    bytes_sent: int
    path: str
    error: Optional[str] = None


@dataclass
class DownloadResult:
    """Result structure for file download operations."""

    success: bool
    request_id_download_url: Optional[str]
    request_id_sync: Optional[str]
    http_status: Optional[int]
    bytes_received: int
    path: str
    local_path: str
    error: Optional[str] = None


class FileChangeEvent:
    """Represents a single file change event."""

    def __init__(
        self,
        event_type: str = "",
        path: str = "",
        path_type: str = "",
    ):
        """
        Initialize a FileChangeEvent.

        Args:
            event_type (str): Type of the file change event (e.g., "modify", "create", "delete").
            path (str): Path of the file or directory that changed.
            path_type (str): Type of the path ("file" or "directory").
        """
        self.event_type = event_type
        self.path = path
        self.path_type = path_type

    def __repr__(self):
        return f"FileChangeEvent(event_type='{self.event_type}', path='{self.path}', path_type='{self.path_type}')"

    def _to_dict(self) -> Dict[str, str]:
        """Convert to dictionary representation."""
        return {
            "eventType": self.event_type,
            "path": self.path,
            "pathType": self.path_type,
        }

    @classmethod
    def _from_dict(cls, data: Dict[str, str]) -> "FileChangeEvent":
        """Create FileChangeEvent from dictionary."""
        return cls(
            event_type=data.get("eventType", ""),
            path=data.get("path", ""),
            path_type=data.get("pathType", ""),
        )


class FileChangeResult(ApiResponse):
    """
    Result of file change detection operations.
    """

    def __init__(
        self,
        request_id: str = "",
        success: bool = False,
        events: Optional[List[FileChangeEvent]] = None,
        raw_data: str = "",
        error_message: str = "",
    ):
        """
        Initialize a FileChangeResult.

        Args:
            request_id (str, optional): Unique identifier for the API request.
                Defaults to "".
            success (bool, optional): Whether the operation was successful.
                Defaults to False.
            events (List[FileChangeEvent], optional): List of file change events.
                Defaults to None.
            raw_data (str, optional): Raw response data for debugging. Defaults to "".
            error_message (str, optional): Error message if the operation failed.
                Defaults to "".
        """
        super().__init__(request_id)
        self.success = success
        self.events = events or []
        self.raw_data = raw_data
        self.error_message = error_message

    def has_changes(self) -> bool:
        """
        Check if there are any file changes.

        Returns:
            bool: True if there are any file change events, False otherwise.
        """
        return len(self.events) > 0

    def get_modified_files(self) -> List[str]:
        """
        Get list of modified file paths.

        Returns:
            List[str]: List of file paths that were modified.
        """
        return [
            event.path
            for event in self.events
            if event.event_type == "modify" and event.path_type == "file"
        ]

    def get_created_files(self) -> List[str]:
        """
        Get list of created file paths.

        Returns:
            List[str]: List of file paths that were created.
        """
        return [
            event.path
            for event in self.events
            if event.event_type == "create" and event.path_type == "file"
        ]

    def get_deleted_files(self) -> List[str]:
        """
        Get list of deleted file paths.

        Returns:
            List[str]: List of file paths that were deleted.
        """
        return [
            event.path
            for event in self.events
            if event.event_type == "delete" and event.path_type == "file"
        ]


class FileInfoResult(ApiResponse):
    """Result of file info operations."""

    def __init__(
        self,
        request_id: str = "",
        success: bool = False,
        file_info: Optional[Dict[str, Any]] = None,
        error_message: str = "",
    ):
        """
        Initialize a FileInfoResult.

        Args:
            request_id (str, optional): Unique identifier for the API request.
                Defaults to "".
            success (bool, optional): Whether the operation was successful.
                Defaults to False.
            file_info (Dict[str, Any], optional): File information. Defaults to None.
            error_message (str, optional): Error message if the operation failed.
                Defaults to "".
        """
        super().__init__(request_id)
        self.success = success
        self.file_info = file_info or {}
        self.error_message = error_message
    
    @property
    def is_file(self) -> bool:
        """Check if the path is a file."""
        return self.file_info.get("isFile", False)
    
    @property
    def is_directory(self) -> bool:
        """Check if the path is a directory."""
        return self.file_info.get("isDirectory", False)
    
    @property
    def size(self) -> int:
        """Get file size in bytes."""
        return self.file_info.get("size", 0)
    
    @property
    def permissions(self) -> str:
        """Get file permissions."""
        return self.file_info.get("permissions", "")


class DirectoryEntry:
    """Wrapper for directory entry with attribute access."""
    
    def __init__(self, entry_dict: Dict[str, Any]):
        self._data = entry_dict
    
    @property
    def name(self) -> str:
        """Get entry name."""
        return self._data.get("name", "")
    
    @property
    def is_file(self) -> bool:
        """Check if entry is a file."""
        # Support both key formats for compatibility
        return self._data.get("isFile", self._data.get("is_file", False))
    
    @property
    def is_directory(self) -> bool:
        """Check if entry is a directory."""
        # Support both key formats for compatibility
        return self._data.get("is_directory", self._data.get("isDirectory", False))
    
    @property
    def size(self) -> int:
        """Get entry size."""
        return self._data.get("size", 0)


class DirectoryListResult(ApiResponse):
    """Result of directory listing operations."""

    def __init__(
        self,
        request_id: str = "",
        success: bool = False,
        entries: Optional[List[Dict[str, Any]]] = None,
        error_message: str = "",
    ):
        """
        Initialize a DirectoryListResult.

        Args:
            request_id (str, optional): Unique identifier for the API request.
                Defaults to "".
            success (bool, optional): Whether the operation was successful.
                Defaults to False.
            entries (List[Dict[str, Any]], optional): Directory entries. Defaults to
                None.
            error_message (str, optional): Error message if the operation failed.
                Defaults to "".
        """
        super().__init__(request_id)
        self.success = success
        self._entries = entries or []
        self.error_message = error_message
    
    @property
    def entries(self) -> List[DirectoryEntry]:
        """Get directory entries with attribute access."""
        return [DirectoryEntry(entry) for entry in self._entries]


class FileContentResult(ApiResponse):
    """Result of file read operations."""

    def __init__(
        self,
        request_id: str = "",
        success: bool = False,
        content: str = "",
        error_message: str = "",
    ):
        """
        Initialize a FileContentResult.

        Args:
            request_id (str, optional): Unique identifier for the API request.
                Defaults to "".
            success (bool, optional): Whether the operation was successful.
                Defaults to False.
            content (str, optional): File content. Defaults to "".
            error_message (str, optional): Error message if the operation failed.
                Defaults to "".
        """
        super().__init__(request_id)
        self.success = success
        self.content = content
        self.error_message = error_message


class MultipleFileContentResult(ApiResponse):
    """Result of multiple file read operations."""

    def __init__(
        self,
        request_id: str = "",
        success: bool = False,
        contents: Optional[Dict[str, str]] = None,
        error_message: str = "",
    ):
        """
        Initialize a MultipleFileContentResult.

        Args:
            request_id (str, optional): Unique identifier for the API request.
                Defaults to "".
            success (bool, optional): Whether the operation was successful.
                Defaults to False.
            contents (Dict[str, str], optional): Dictionary of file paths to file
                contents. Defaults to None.
            error_message (str, optional): Error message if the operation failed.
                Defaults to "".
        """
        super().__init__(request_id)
        self.success = success
        self.contents = contents or {}
        self.error_message = error_message


class FileSearchResult(ApiResponse):
    """Result of file search operations."""

    def __init__(
        self,
        request_id: str = "",
        success: bool = False,
        matches: Optional[List[str]] = None,
        error_message: str = "",
    ):
        """
        Initialize a FileSearchResult.

        Args:
            request_id (str, optional): Unique identifier for the API request.
                Defaults to "".
            success (bool, optional): Whether the operation was successful.
                Defaults to False.
            matches (List[str], optional): Matching file paths. Defaults to None.
            error_message (str, optional): Error message if the operation failed.
                Defaults to "".
        """
        super().__init__(request_id)
        self.success = success
        self.matches = matches or []
        self.error_message = error_message

